velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,356 @@
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator>=(const PrefixSortIterator& other) const {
+    return prefix_ >= other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator<=(const PrefixSortIterator& other) const {
+    return prefix_ <= other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator==(const PrefixSortIterator& other) const {
+    return prefix_ == other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator!=(const PrefixSortIterator& other) const {
+    return prefix_ != other.prefix_;
+  }
+
+ private:
+  char* prefix_;
+  const uint64_t entrySize_;
+};
+} // namespace detail
+
+/// Provides a series of methods (mostly required by the sort
+/// algorithm) for PrefixSort. Maintains the buffer for swap operations and
+/// some necessary context information such as entrySize.
+class PrefixSortContext {
+ public:
+  /// @param swapBuffer The buffer must be at least entrySize bytes long.
+  PrefixSortContext(uint64_t entrySize, char* swapBuffer)
+      : entrySize_(entrySize), swapBuffer_(swapBuffer) {
+    VELOX_CHECK_NOT_NULL(swapBuffer_);
+  }
+
+  // Within quickSort, when the input data length < kSmallSort, use insert-sort
+  // in place of quick-sort.
+  // With finding a central element, there are extra comparisons. While
+  // this is cheap for large arrays, it is expensive for small arrays.
+  // Therefore, chooses the middle element of smaller arrays (=kSmallSort),
+  // the median of the first, middle and last elements of a mid-sized array
+  // (> kSmallSort and < kMediumSort),
+  // and the pseudo-median of nine evenly spaced elements of a large
+  // array(> kMediumSort).
+  // The two threshold values are given in the reference paper "Engineering a
+  // Sort Function".
+  static const int kSmallSort = 7;
+  static const int kMediumSort = 40;
+
+  template <typename TCompare>
+  void quickSort(char* start, char* end, TCompare compare) const {
+    quickSort(
+        detail::PrefixSortIterator(start, entrySize_),
+        detail::PrefixSortIterator(end, entrySize_),
+        compare);
+  }
+
+  /// For testing only.
+  template <typename TCompare>
+  FOLLY_ALWAYS_INLINE static char* testingMedian3(
+      char* a,
+      char* b,
+      char* c,
+      const uint64_t entrySize,
+      TCompare cmp) {
+    return *median3(
+        detail::PrefixSortIterator(a, entrySize),
+        detail::PrefixSortIterator(b, entrySize),
+        detail::PrefixSortIterator(c, entrySize),
+        cmp);
+  }
+
+ private:
+  FOLLY_ALWAYS_INLINE void swap(
+      const detail::PrefixSortIterator& lhs,
+      const detail::PrefixSortIterator& rhs) const {
+    simd::memcpy(swapBuffer_, *lhs, entrySize_);
+    simd::memcpy(*lhs, *rhs, entrySize_);
+    simd::memcpy(*rhs, swapBuffer_, entrySize_);
+  }
+
+  FOLLY_ALWAYS_INLINE void rangeSwap(
+      const detail::PrefixSortIterator& start1,
+      const detail::PrefixSortIterator& start2,
+      uint64_t length) const {
+    for (uint64_t i = 0; i < length; i++) {
```

Comment:
If we can afford a larger `swapBuffer_`, can we coalesce these copies to make it faster?

Reply:
I do not do this because this :
1. The entrySize is not fixed  for all querys,  it is difficult to decide a suitable batchSize for all case. We do this in our internal project, because we can ajust the batchSize through query-config / sql hint,  but I think it is not a good way for community.
2.The rangeSwap method seems cost far less than compare & swap method,  the benefits are not big, but the code is more complex.

Reply:
OK (2) sounds a good reason not doing it for now